src/psfmachine/tpf.py

```diff
@@ -283,6 +294,130 @@ def plot_tpf(self, tdx):
+            # *4 to convert from pixel to arcseconds
+            self.rmin = int(tab.loc[channel, (str(quarter), "rmin")] * 4)
+            self.rmax = int(tab.loc[channel, (str(quarter), "rmax")] * 4)
+            self.cut_r = int(6 * 4)
+            self.psf_w = tab.loc[channel, str(quarter)].iloc[4:].values
+            # self.psf_w_err = tab.loc[channel, str(quarter)].iloc[4:].values
+            del tab
+        # if channel or quarter data is not in file, then a KeyError is raised
+        except KeyError:
+            raise IOError(
+                "Quarter %i and channel %i has no PRF model data" % (quarter, channel)
+            )
+
+        # create source mask and uncontaminated pixel mask
+        self._get_source_mask()
+        self._get_uncontaminated_pixel_mask()
+
+        # create mean model, but PRF shapes from FFI are in pixels! and TPFMachine
+        # work in arcseconds
+        self._get_mean_model()
+
+        # Re-estimate source flux
+        # -----
+        prior_mu = self.source_flux_estimates
+        prior_sigma = (
+            np.ones(self.mean_model.shape[0]) * 10 * self.source_flux_estimates
+        )
+
+        f, fe = (self.flux).mean(axis=0), ((self.flux_err ** 2).sum(axis=0) ** 0.5) / (
+            self.nt
+        )
+
+        X = self.mean_model.copy()
+        X = X.T
+
+        sigma_w_inv = X.T.dot(X.multiply(1 / fe[:, None] ** 2)).toarray()
+        sigma_w_inv += np.diag(1 / (prior_sigma ** 2))
+        B = X.T.dot((f / fe ** 2))
+        B += prior_mu / (prior_sigma ** 2)
+        ws = np.linalg.solve(sigma_w_inv, B)
+        werrs = np.linalg.inv(sigma_w_inv).diagonal() ** 0.5
+
+        # -----
+
+        # Rebuild source mask
+        ok = np.abs(ws - self.source_flux_estimates) / werrs > 3
+        ok &= ((ws / self.source_flux_estimates) < 10) & (
+            (self.source_flux_estimates / ws) < 10
+        )
+        ok &= ws > 10
+        ok &= werrs > 0
+
+        self.source_flux_estimates[ok] = ws[ok]
+
+        self.source_mask = (
+            self.mean_model.multiply(
+                self.mean_model.T.dot(self.source_flux_estimates)
+            ).tocsr()
+            > 1
+        )
+
+        # Recreate uncontaminated mask
+        self._get_uncontaminated_pixel_mask()
+
+        # Recreate mean model!
+        self._get_mean_model()
+
+        if plot:
+            return self.plot_shape_model()
+        return
+
+    def save_shape_model(self, output=None):
+        """Saves the weights of a PRF fit to a file
+        Parameters
+        ----------
+        output : str, None
+            Output file name. If None, one will be generated.
+        """
+        # asign a file name
+        if output is None:
+            output = "%s/data/shape_model_ch%02i_q%02i.csv" % (
+                PACKAGEDIR,
+                self.tpf_meta["channel"][0],
+                self.tpf_meta["quarter"][0],
+            )
```

Comment:
Is it possible to add a mission keyword here? I don't know if TPFMachine carries that meta data around but I think it might

I think we should make a separate directory for TPFMachine and future FFIMachine

*These ones should be automatically loaded when we use `load_shape_model`*
`data/PRFMachine/kepler_q01_ch01_shape_model.csv`
*These ones are more risky, they depend on how many TPFs were used.*
`data/TPFMachine/kepler_q01_ch01_shape_model.csv`

We may even prefer to have `save` automatically dump the CSV file in the current working directory, rather than storing them in the package.